### PR TITLE
Fix and add test for 1453

### PR DESCRIPTION
**Pull request by charleskawczynski**
This PR fixes and adds a test for #1453.

### CODE REVIEW

**Review by dennisYatunin — APPROVED** [review]
This looks good! I think you can also delete all the methods that specialize on `::Number` in this file, e.g.,
```
rmul(w::Number, X) = rmap(x -> w * x, X)
rmul(X, w::Number) = rmap(x -> x * w, X)
rmul(w1::Number, w2::Number) = w1 * w2
```

**Comment by charleskawczynski** [comment]
> This looks good! I think you can also delete all the methods that specialize on `::Number` in this file, e.g.,

I'm going to leave them there, just to be overly cautious and make sure that this is not a breaking change

**Review by simonbyrne — APPROVED** [review]

**Comment by charleskawczynski** [comment]
I had to fix some method ambiguities, and I'm actually not 100% what behavior we want in these edge cases. cc @simonbyrne. For example:

For 2-argument empty tuple cases, do we want:
```julia
rmap(fn::F, X::Tuple{}, Y::Tuple) where {F} =
    (rmap(fn, first(Y)), rmap(fn, Base.tail(Y))...)
```
or
```julia
rmap(fn::F, X::Tuple{}, Y::Tuple) where {F} = ()
```
or
```julia
rmap(fn::F, X::Tuple{}, Y::Tuple) where {F} = error("oops")
```
?

For now I put the first option, but arguably this should be empty or error.

**Comment by charleskawczynski** [comment]
I also introduced a new function, `nt_names` so that we can specialize on that, and remove the forced specialization on all the `rmap` methods. This way, forced specialization on `nt_names` can be reused, potentially resulting in fewer specializations at the cost of one more method.

File: test/aqua.jl [hunk 1]
```diff
@@ -22,7 +22,7 @@ using Aqua
     for method_ambiguity in ambs
         @show method_ambiguity
     end
-    @test length(ambs) ≤ 16
+    @test length(ambs) ≤ 13
```
Comment (simonbyrne) [inline test/aqua.jl:25]: this is failing https://github.com/CliMA/ClimaCore.jl/actions/runs/6214646588/job/16866787715?pr=1454#step:7:686
```
Aqua tests (performance): Test Failed at /home/runner/work/ClimaCore.jl/ClimaCore.jl/test/aqua.jl:25
  Expression: length(ambs) ≤ 13
   Evaluated: 15 ≤ 13
```

Comment (charleskawczynski) [reply test/aqua.jl:25]: Ah, interesting, it was passing for me locally. I can bump it back up. @simonbyrne, can you re-review the functionality aspect I introduced here?

**Pull request merged**